kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkTask.java

```diff
@@ -275,13 +297,30 @@ public Map<TopicPartition, OffsetAndMetadata> preCommit(
       Map<TopicPartition, OffsetAndMetadata> offsets
   ) {
     Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = new HashMap<>();
+    Map<TopicPartition, Set<String>> filesToCommit = new HashMap<>();
     for (TopicPartition tp : topicPartitionWriters.keySet()) {
       Long offset = topicPartitionWriters.get(tp).getOffsetToCommitAndReset();
+      Set<String> files = topicPartitionWriters.get(tp).getCommitFilesNames();
       if (offset != null) {
         log.trace("Forwarding to framework request to commit offset: {} for {}", offset, tp);
         offsetsToCommit.put(tp, new OffsetAndMetadata(offset));
       }
+      if (!files.isEmpty()) {
```

Comment:
There is a data lose 

E.g. rotation by size 3
put 3 messages -> rotation -> s3 sent ok -> failed to produce to kafka -> do not commit offset
put 3 messages -> rotation -> s3 sent failed -> start over (failed files are cleared so lost)

Reply:
Your point is valid. yet with a small change, we can possibly avoid this data loss. 

When we start over we start either from the latest committed offset in case of a total crash or from the minimum of `currentStartOffset` (`io/confluent/connect/s3/S3SinkTask.java:259`) in case of `RetriableException`.

1. In case of a total crash we are good as we start from the latest committed offset and according to your example we didn't commit anything back, so we will produce all again.
2. In case we have `RetriableException` we will seek offset to the minimum of `currentStartOffset`  (`io/confluent/connect/s3/S3SinkTask.java:259`) and create a new `TopicPartitionWriter` and thus lose the previous files inside `commitFilesNames` (here is the data loss you pointed). In this case, we need to add them back when we create the new instance (not so good) or roll back to the latest committed offset which might be < `min(currentStartOffset)` (wasteful).

Let's further discuss this

Reply:
Another idea is to move `commitFilesNames ` to `S3SinkTask` and allow the `precommit` method to commit only if that data structure is empty, i.e. all produced to kafka